src/Controller/AppController.php

```diff
@@ -332,4 +332,22 @@ protected function getObjectNav($id): array
 
         return (array)Hash::get($objectNav, sprintf('%s.%s', $objectNavModule, $id), []);
     }
+
+    /**
+     * Set session data for `failedSave.{type}.{id}` and `failedSave.{type}.{id}__timestamp`.
+     *
+     * @param string $type The object type.
+     * @param array $data The data to store into session.
+     * @return void
+     */
+    protected function setDataFromFailedSave($type, $data): void
```

Comment:
move also to `ModulesComponent` ? 
is probably convenient to have both methods writing and reading this data in session in the same class